File: frontend/src/services/coding.ts
import { fetchWithAuth } from './api';

export interface TestCase {
    test_case_id: string; // UUID
    input: string;
    expected_output: string;
    is_sample: boolean;
    points: number;
}

export interface Problem {
    problem_id: string; // UUID
    problem_code: string;
    title: string;
    description?: string;
    difficulty: string;
    time_limit: number;
    memory_limit: number;
    allowed_languages: string[];
    sample_test_cases?: TestCase[];
    course_id?: string;
}

export interface Submission {
    submission_id: string; // UUID
    student_id: string; // UUID
    problem_id: string; // UUID
    language: string;
    source_code: string;
    status: string;
    score: number;
    execution_time?: number;
    memory_used?: number;
    test_cases_passed: number;
    total_test_cases: number;
    created_at: string;
    test_results?: any[];
}

export interface StudentBasicInfo {
    student_id: string;
    student_code: string;
    user: {
        full_name: string;
        email: string;
    };
}

export interface SubmissionWithStudent extends Submission {
    student: StudentBasicInfo;
}

export interface SubmissionCreate {
    problem_id: string; // UUID
    language: string;
    source_code: string;
    assignment_id?: string; // UUID
}

export const codingService = {
    async getProblems(courseId?: string) {
        const url = courseId ? `/coding/problems?course_id=${courseId}` : '/coding/problems';
        const response = await fetchWithAuth(url);
        if (!response.ok) throw new Error('Failed to fetch problems');
        return response.json();
    },

    async getProblem(id: string) {
        const response = await fetchWithAuth(`/coding/problems/${id}`);
        if (!response.ok) throw new Error('Failed to fetch problem');
        return response.json();
    },

    async createSubmission(data: SubmissionCreate) {
        const response = await fetchWithAuth('/coding/submissions', {
            method: 'POST',
            body: JSON.stringify(data)
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Failed to submit solution');
        }
        return response.json();
    },

    async getSubmission(id: string) {
        const response = await fetchWithAuth(`/coding/submissions/${id}`);
        if (!response.ok) throw new Error('Failed to fetch submission');
        return response.json();
    },

    async getMySubmissions() {
        const response = await fetchWithAuth('/coding/my-submissions');
        if (!response.ok) throw new Error('Failed to fetch submissions');
        return response.json();
    },

    async getProblemSubmissions(problemId: string) {
        const response = await fetchWithAuth(`/coding/problems/${problemId}/submissions`);
        if (!response.ok) throw new Error('Không thể tải lịch sử nộp bài');
        return response.json();
    },

    async getAllProblemSubmissions(problemId: string) {
        const response = await fetchWithAuth(`/coding/problems/${problemId}/all-submissions`);
        if (!response.ok) throw new Error('Không thể tải danh sách nộp bài');
        return response.json();
    },

    // --- Admin Functions ---

    async createProblem(data: Omit<Problem, 'problem_id' | 'sample_test_cases'>) {
        const response = await fetchWithAuth('/coding/problems', {
            method: 'POST',
            body: JSON.stringify(data)
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Không thể tạo bài tập');
        }
        return response.json();
    },

    async updateProblem(id: string, data: Omit<Problem, 'problem_id' | 'sample_test_cases'>) {
        const response = await fetchWithAuth(`/coding/problems/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Không thể cập nhật bài tập');
        }
        return response.json();
    },

    async deleteProblem(id: string) {
        const response = await fetchWithAuth(`/coding/problems/${id}`, {
            method: 'DELETE'
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Không thể xóa bài tập');
        }
        return response.json();
    },

    async getTestCases(problemId: string) {
        const response = await fetchWithAuth(`/coding/problems/${problemId}/test-cases`);
        if (!response.ok) throw new Error('Không thể tải test cases');
        return response.json();
    },

    async createTestCase(problemId: string, data: Omit<TestCase, 'test_case_id'>) {
        const response = await fetchWithAuth(`/coding/problems/${problemId}/test-cases`, {
            method: 'POST',
            body: JSON.stringify(data)
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Không thể tạo test case');
        }
        return response.json();
    },

    async deleteTestCase(testCaseId: string) {
        const response = await fetchWithAuth(`/coding/test-cases/${testCaseId}`, {
            method: 'DELETE'
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Không thể xóa test case');
        }
        return response.json();
    },

    async updateTestCase(testCaseId: string, data: Omit<TestCase, 'test_case_id'>) {
        const response = await fetchWithAuth(`/coding/test-cases/${testCaseId}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Không thể cập nhật test case');
        }
        return response.json();
    },

    async bulkCreateTestCases(problemId: string, testCases: Omit<TestCase, 'test_case_id'>[]) {
        // Create multiple test cases sequentially
        const results = [];
        for (const tc of testCases) {
            const response = await fetchWithAuth(`/coding/problems/${problemId}/test-cases`, {
                method: 'POST',
                body: JSON.stringify(tc)
            });
            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.detail || `Không thể tạo test case: ${tc.input.substring(0, 20)}...`);
            }
            results.push(await response.json());
        }
        return results;
    }
};

File: fastapi/app/api/v1/endpoints/coding.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from typing import List
from uuid import UUID

from app.db.database import get_db
from app.schemas import coding as schemas
from app.api.v1.endpoints.users import get_current_user
from app.services.problem_service import ProblemService
from app.services.submission_service import SubmissionService

router = APIRouter()

# --- Problem Endpoints ---

@router.get("/problems", response_model=List[schemas.ProblemList])
def read_problems(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    service = ProblemService(db)
    return service.get_problems(skip, limit)

@router.get("/problems/{problem_id}", response_model=schemas.Problem)
def read_problem(problem_id: UUID, db: Session = Depends(get_db)):
    service = ProblemService(db)
    return service.get_problem(problem_id)

@router.post("/problems", response_model=schemas.Problem)
def create_problem(
    problem: schemas.ProblemCreate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Create a new problem (admin/lecturer only)"""
    service = ProblemService(db)
    return service.create_problem(problem, current_user)

@router.put("/problems/{problem_id}", response_model=schemas.Problem)
def update_problem(
    problem_id: UUID,
    problem: schemas.ProblemCreate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Update an existing problem"""
    service = ProblemService(db)
    return service.update_problem(problem_id, problem, current_user)

@router.delete("/problems/{problem_id}")
def delete_problem(
    problem_id: UUID,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Delete a problem"""
    service = ProblemService(db)
    return service.delete_problem(problem_id, current_user)


# --- Test Case Endpoints ---

@router.post("/problems/{problem_id}/test-cases", response_model=schemas.TestCase)
def create_test_case(
    problem_id: UUID,
    test_case: schemas.TestCaseCreate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Add a test case to a problem"""
    service = ProblemService(db)
    return service.create_test_case(problem_id, test_case, current_user)

@router.get("/problems/{problem_id}/test-cases", response_model=List[schemas.TestCase])
def get_test_cases(
    problem_id: UUID,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get all test cases for a problem (admin/lecturer only)"""
    service = ProblemService(db)
    return service.get_test_cases(problem_id, current_user)

@router.delete("/test-cases/{test_case_id}")
def delete_test_case(
    test_case_id: UUID,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Delete a test case"""
    service = ProblemService(db)
    return service.delete_test_case(test_case_id, current_user)


# --- Submission Endpoints ---

@router.post("/submissions", response_model=schemas.Submission)
def create_submission(
    submission: schemas.SubmissionCreate, 
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    service = SubmissionService(db)
    return service.create_submission(submission, current_user)

@router.get("/submissions/{submission_id}", response_model=schemas.Submission)
def read_submission(submission_id: UUID, db: Session = Depends(get_db)):
    service = SubmissionService(db)
    return service.get_submission(submission_id)

@router.get("/my-submissions", response_model=List[schemas.Submission])
def read_my_submissions(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    service = SubmissionService(db)
    return service.get_my_submissions(current_user)

@router.get("/problems/{problem_id}/submissions", response_model=List[schemas.SubmissionListItem])
def read_problem_submissions(
    problem_id: UUID,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Get current user's submissions for a specific problem"""
    service = SubmissionService(db)
    return service.get_problem_submissions(problem_id, current_user)
